[RDY] Run builtin TUI in a another thread

tarruda opened the pull request:
The goal is to refactor the builtin TUI allow fixing issues such as #2093

tarruda committed:
tui: move term_input.inl into the input.c module

tarruda committed:
tui: Move screen state tracking to new "ugrid" module

The ugrid module implements a unicode "drawing" grid and is used to store TUI
screen state. Later this module will be reused in other layers.

tarruda committed:
contrib: Implement gdb pretty printer for UGrid

oni-link commented on src/nvim/ugrid.c:
```diff
@@ -0,0 +1,137 @@
+
+void ugrid_free(UGrid *grid)
+{
+  xfree(grid->cells);
```
What about the different grid rows? Don't we need to call `destroy_cells(grid);` here?

tarruda replied:
:+1:

oni-link commented on src/nvim/tui/input.c:
```diff
@@ -1,21 +1,64 @@
+  stream_close(&input->read_stream, NULL);
+  termkey_destroy(input->tk);
+  // Run once to remove references to input/timer handles
+  loop_poll_events(&loop, 0);
```
If the libuv loop variable `stop_flag` was set (by `loop_on_put()`), `uv_run()` would return without removing any references. Do we need to call this function twice?

tarruda committed:
travis: Enable TSAN for sanity checking the terminal UI

tarruda commented:
@aeruder can you test if this branch fixes #2093?

justinmk commented:
Can you explain the design here? What is ui_bridge? 

I have also wondered why do we make a disctinction between "UI" and non-UI clients? Why isn't it all unified under API, i.e. UI messages are just a API messages that a client may choose to respond to.

tarruda commented:
> Can you explain the design here? What is ui_bridge?

It's not ready yet(thus missing documentation), but it simply registers a surrogate UI that forwards messages to the real U(running in another thread). ui_bridge.c and ugrid.c are not tied to terminal and can be reused in another builtin UI implementation. I borrowed the concept from the pygtk UI: https://github.com/neovim/python-client/blob/master/neovim/ui/ui_bridge.py

> I have also wondered why do we make a disctinction between "UI" and non-UI clients? Why isn't it all unified under API, i.e. UI messages are just a API messages that a client may choose to respond to.

Basically the only difference between UI and non-UI clients is that UI clients will call `ui_attach` which registers a UI(allocates a UI structure with height/width) and subscribes to UI notifications. Note that the UI layer is actually independent from msgpack-rpc, which is only used for remote UIs, so when we say "UI client" we actually mean "remote UI client".

justinmk commented:
> UI layer is actually independent from msgpack-rpc, which is only used for remote UIs, so when we say "UI client" we actually mean "remote UI client".

By "UI layer" here do you mean the TUI? It seems to me that `nvim/ui.c` is sending RPC messages (except in the case of TUI) via `UI_CALL`. It would be good for comprehension if these messages were sent over the same mechanism used by non-UI API methods. I mention this now because I don't want to wait until it is too late. 

Current state of Neovim API:

- We have "UI messages"
- API publish/subscribe (not used by anything yet, AFAIK)
- API methods
- another direction called "smart UI"

This does not seem ideal. The code is well-structured and clean, but from a _concept_ perspective, I am looking for opportunities to unify our API into a single concept. In other words, why do we have a message system for only _UI_ messages? Why not have a generalized message system for use by any client?

Does this make sense?

(Aside: I will repeat here that I do not like the idea of delegating API publish/subscribe to vimscript, because it makes `nvim` API dependent on the runtime files.)

tarruda commented:
> By "UI layer" here do you mean the TUI? It seems to me that nvim/ui.c is sending RPC messages (except in the case of TUI) via UI_CALL.

nvim/ui.c is simply calling handler functions of registered UI structs, this is not related to RPC because the UI structs like in the same process. Before this PR there are two modules that implement UI struct handlers:

- tui.c where the handler functions generate escape codes for drawing the UI on the terminal
- remote_ui.c where the handler functions generate msgpack-rpc notifications for clients that called `ui_attach`

>  It would be good for comprehension if these messages were sent over the same mechanism used by non-UI API methods

I'm not sure I follow completely, do you mean we should use msgpack-rpc for communicating between modules in the same process or that we move TUI to another process?

If you mean the latter, this is something I had considered before the project started(may even be visible in the pre-fundraiser documentation): nvim would split vim into client and server programs where the client implements the UI and the server the core logic.

I gave up on that idea because I saw it would result in more work and managed to structure the UI layer in a way that allowed builtin UIs without compromising the remote UI feature.

> Current state of Neovim API:

> We have "UI messages"

As explained above, ui.c is not part of the API, the remote_ui.c module exposes msgpack-rpc handlers for ui_attach/ui_resize/ui_detach for msgpack-rpc clients that want to implement a UI in another process, but the ui.c module itself only sees an in-process UI.

> API publish/subscribe (not used by anything yet, AFAIK)

True. The reason I want to remove this and implement in vimscript is because the whole range of editor events is already exposed to vimscript, and the amount of vimscript code to make this available to plugins is significantly less than implementing it in C .

> This does not seem ideal. The code is well-structured and clean, but from a concept perspective, I am looking for opportunities to unify our API into a single concept. In other words, why do we have a message system for only UI messages? Why not have a generalized message system for use by any client?

Again, we don't have UI messages. We have a UI struct that contains a bunch of function pointers, and this is exposed to msgpack-rpc by remote_ui.c. The only other way I see to generalize the UI concept behind msgpack-rpc is to externalize the TUI into another program(at which point we'd be distributing two binaries: one server and one client.

> (Aside: I will repeat here that I do not like the idea of delegating API publish/subscribe to vimscript, because it makes nvim API dependent on the runtime files.)

IMO this is not enough reason to implement it in C which would result in a lot more work, especially because we can embed essential vimscript into the executable(like it was done for `:terminal` for example). It should be trivial to modify the build system to merge vimscript with C before compilation.

oni-link commented on src/nvim/tui/tui.c:
```diff
@@ -202,18 +126,94 @@ static void tui_stop(UI *ui)
   unibi_destroy(data->ut);
-  destroy_screen(data);
+  ugrid_free(&data->grid);
+  data->stop = true;
```
This "stop signal" does not reaches `tui_main()` before `fg` is used, because `tui_main()` should still be polling for input.

tarruda replied:
Not sure I follow, what does `fg` mean in this context?' When `tui_stop` is called by the ui thread, the call stack should look like this:

- `tui_main`
- `loop_poll_events`
- ...
- `tui_stop`


so when `loop_poll_events` returns the while loop in `tui_main` will exit

oni-link replied:
If I use `<CTRL-z>`, `tui_stop()` is called and I expected that `tui_main()` would exit at once (thread ends).
For example, I don't see the message with the following patch until I use `fg`:
```diff
diff --git a/src/nvim/tui/tui.c b/src/nvim/tui/tui.c
index f678428..63a359e 100644
--- a/src/nvim/tui/tui.c
+++ b/src/nvim/tui/tui.c
@@ -189,6 +189,7 @@ static void tui_main(UIBridgeData *bridge, UI *ui)
   while (!data->stop) {
     loop_poll_events(&tui_loop, -1);
   }
+  fprintf(stderr,"tui_main ends\n");
 
   loop_close(&tui_loop);
   xfree(data);
```

Also program crashes when `fg` is used (here patch is not included).
```c
Program terminated with signal SIGABRT, Aborted.
#0  0x00007f744a839a38 in raise () from /lib64/libc.so.6
(gdb) bt
#0  0x00007f744a839a38 in raise () from /lib64/libc.so.6
#1  0x00007f744a83ae0a in abort () from /lib64/libc.so.6
#2  0x000000000062abc7 in uv_mutex_lock (mutex=<optimized out>)
    at /home/oni-link/git/neovim/.deps/build/src/libuv/src/unix/thread.c:120
#3  0x000000000061d0bb in loop_schedule (loop=0x7f7449c528c0, event=...) at /home/oni-link/git/neovim/src/nvim/event/loop.c:62
#4  0x0000000000618890 in tui_scheduler (event=..., d=0x7f744a03a2c0) at /home/oni-link/git/neovim/src/nvim/tui/tui.c:202
#5  0x00000000004d8a8d in ui_bridge_put (b=0x7f744a024a80, text=0x7ffc4f5af790 " ", size=1)
    at /home/oni-link/git/neovim/src/nvim/ui_bridge.c:244
#6  0x00000000005d0f44 in send_output (ptr=0x7ffc4f5af760) at /home/oni-link/git/neovim/src/nvim/ui.c:353
#7  0x00000000005d0d46 in ui_puts (str=0x7ffc4f5af790 " ") at /home/oni-link/git/neovim/src/nvim/ui.c:306
#8  0x00000000005d0da0 in ui_putc (c=32 ' ') at /home/oni-link/git/neovim/src/nvim/ui.c:314
#9  0x000000000049beba in screen_char (off=1223, row=8, col=135) at /home/oni-link/git/neovim/src/nvim/screen.c:5732
#10 0x000000000049c3c4 in screen_fill (start_row=1, end_row=68, start_col=0, end_col=136, c1=126, c2=32, attr=6)
    at /home/oni-link/git/neovim/src/nvim/screen.c:5865
#11 0x0000000000490b2b in win_draw_end (wp=0x7f744a04e000, c1=126, c2=32, row=1, endrow=68, hl=HLF_EOB)
    at /home/oni-link/git/neovim/src/nvim/screen.c:1648
#12 0x00000000004904c1 in win_update (wp=0x7f744a04e000) at /home/oni-link/git/neovim/src/nvim/screen.c:1527
#13 0x000000000048d5d4 in update_screen (type=40) at /home/oni-link/git/neovim/src/nvim/screen.c:419
#14 0x000000000049f972 in screen_resize (width=136, height=70) at /home/oni-link/git/neovim/src/nvim/screen.c:7279
#15 0x00000000005d04fd in ui_refresh () at /home/oni-link/git/neovim/src/nvim/ui.c:142
#16 0x00000000005ba2dc in ex_stop (eap=0x7ffc4f5afb60) at /home/oni-link/git/neovim/src/nvim/ex_docmd.c:5937
#17 0x00000000005b1a5c in do_one_cmd (cmdlinep=0x7ffc4f5afc80, sourcing=1, cstack=0x7ffc4f5afd90, fgetline=0x0, cookie=0x0)
    at /home/oni-link/git/neovim/src/nvim/ex_docmd.c:2195
#18 0x00000000005ae5df in do_cmdline (cmdline=0x69fd5e "st", fgetline=0x0, cookie=0x0, flags=11)
    at /home/oni-link/git/neovim/src/nvim/ex_docmd.c:621
#19 0x00000000005adc5d in do_cmdline_cmd (cmd=0x69fd5e "st") at /home/oni-link/git/neovim/src/nvim/ex_docmd.c:293
#20 0x00000000005661ee in nv_suspend (cap=0x7ffc4f5b0300) at /home/oni-link/git/neovim/src/nvim/normal.c:6011
#21 0x000000000055a9b5 in normal_cmd (oap=0x7ffc4f5b03a0, toplevel=true) at /home/oni-link/git/neovim/src/nvim/normal.c:912
#22 0x0000000000553227 in main_loop (cmdwin=0, noexmode=0) at /home/oni-link/git/neovim/src/nvim/main.c:747
#23 0x0000000000552be0 in main (argc=3, argv=0x7ffc4f5b0668) at /home/oni-link/git/neovim/src/nvim/main.c:534
```

tarruda replied:
:+1:

tarruda replied:
should be fixed now

tarruda committed:
build: Disable gcov on TSAN builds

It seems GCOV does not play well with TSAN as it reports false data races

justinmk commented:
Will this PR also fix the problem where operator-pending is canceled by (I assume) an event loop run? E.g., try this:

```
:te top
<c-\><c-n>
<c-w>                <--this enters operator-pending mode
```

In the last step, the operator-pending mode is canceled by the next refresh initialized by `top`.

I suspect this is just a simple regression, because `<c-\>` operator is not interrupted by `top`, nor is command-line mode.

torpak commented:
Does this set the terminal to 8-bit mode?

tarruda commented:
> Will this PR also fix the problem where operator-pending is canceled by (I assume) an event loop run? 

No. To fix this problem, we need to handle `K_EVENT` in every function that reads another character for the purpose of finishing an operator pending mode. It's not an absurd amount of work, but it requires a search through normal.c and ops.c

> Does this set the terminal to 8-bit mode?

@torpak no

torpak commented:
@tarruda From the Chatlog in #2093 it seemed, that setting the terminal to 8-bit mode and modifying i/o code to cope with that, was the only clean fix for the <ESC> problem.
Is this wrong, or would the clean fix just be to much work?

tarruda commented:
> @tarruda From the Chatlog in #2093 it seemed, that setting the terminal to 8-bit mode and modifying i/o code to cope with that, was the only clean fix for the problem.
Is this wrong, or would the clean fix just be to much work?

8-bit mode does allow to unambiguously distinguish <kbd>alt</kbd> combinations from the <kbd>esc</kbd> key, but it also incompatible with unicode(since all bits are used we can't parse the higher utf-8 sequences), so it's not really a fix.

I have explained the problem in a [comment](https://github.com/neovim/neovim/issues/2093#issuecomment-78386448):

> The problem reported by @aeruder is the same, but the reason it is not fixed by lowering the timeout is because showmatch uses a sleep function which causes a lag that results in esc+j arriving together in the input buffer.

So my goal here was to remove the lag factor caused by showmatch's sleep, but the only way to completely fix the ambiguity between alt combos and esc is change the key encoding used by both terminals and terminal programs, which is one of @leonerd goals with libtermkey. AFAIK libtermkey does not yet allow for `\x1b` to be unambiguously interpreted as the esc key because it still uses the old encoding for alt modifier even though it supports a different encoding for other key combinations(@leonerd please correct me if I'm wrong)

leonerd commented:
@tarruda Kindof.

A program could do better than the current logic on some terminals, by asking for pure 8bit mode (S8C1T) and then immediately querying the mode again. If the result comes back as an 8bit single-byte CSI, then it can presume the mode setting was successful, and now the ESC prefix byte won't be seen in multibyte sequences; only as an Alt- prefix or a real Escape key. On such a terminal, it could therefore avoid needing to use that waiting timeout. If it doesn't receive such a response, then it just has to fall back to the current behaviour. So it's no worse than now, but on a cooperating terminal it would be better.

As it stands, the logic of the existing libtermkey isn't set up to be talking out *to* the terminal, only receiving responses from it. This is the primary motivation in my goal to merge it in with libtickit, where it would be able to perform such an interaction.

As to the interaction with UTF-8, you're half-right. A strict interpretation of both the UTF-8 and the ECMA-35 encoding schemes would suggest they are incompatible. However, a more careful observation would note that any of the high 8-bit bytes used by ECMA-35 in pure 8-bit mode are only used by UTF-8 as "continuation" bytes, never as the first byte of a valid UTF-8 multi-byte character. A bit of detection in the decoder can therefore pull these two streams apart, and indeed is what libtermkey (and hence libtickit) do.

leonerd commented:
Specifically with respect to your use of the `sleep` function, I really really would advise not doing that. With the ongoing ambiguity of Escape-then-letter as compared Alt-letter (even with S8C1T mode active) it is imperative that libtermkey (or libtickit) get as low-latency input as possible, in order to minimise the chance of a false read of key presses due to conflated bytes that should have arrived separately.

Ideally, the inner core of a program would be written in a purely non-blocking style, using a central `select()` or similar loop to drive libtermkey/etc.. as soon as input is available. I appreciate however that turning an existing large codebase like (neo)vim inside-out to such an arrangement would be difficult. In such a case, you might find it useful to replace the `sleep` call with a call to small IO-based function that uses calls to libtermkey's input waiting function with a timeout applied, using that to wait until the allotted time, but at the same time pulling key events out of libtermkey and placing them into neovim's input queue. In this way, if a lone ESC byte appears indicating an Escape key, that will come out into the queue correctly before the sleep time is over, thus ensuring it isn't falsely conflated to being an Alt-modified keypress when the next letter arrives.

tarruda commented:
@leonerd Thanks for the info. As for your suggestion:

> Ideally, the inner core of a program would be written in a purely non-blocking style, using a central select() or similar loop to drive libtermkey/etc.. as soon as input is available

This is basically what this PR is aiming to achieve, except in this case the event loop that runs libtermkey is being moved to another thread so it can parse and enqueue keys as soon as data is available, minimizing lag impact when parsing alt sequences

tarruda committed:
log: Make logging thread-safe

tarruda committed:
queue: Allow direct appending to parent queues

tarruda committed:
loop: Improvements for thread-safety

- Implement `loop_schedule` method for queueing events from other threads
- Make `loop_poll_events` `recursive` static variable a field of the Loop
  structure

tarruda committed:
ui: Implement module for thread-safe communication with U

The ui_bridge.c module implements a surrogate UI that forwards calls to another
thread.

tarruda committed:
tui: Refactor for running in another thread

bfredl commented:
I'll be sure to take a look. Note that #2929 will simplify the TUI logic around encoding, I it is more-or-less ready for merge but I kinda forgot about it, I will mark it RDY and merge if no more comments. Sorry if caused extra work for you in the latest refactor here.

tarruda commented:
> Sorry if caused extra work for you in the latest refactor here.

Not at all, separating UI-specific configuration from normal vim options is something I've been meaning to do for a time.

justinmk commented:
I would definitely like to review this later today.

> set_option callback

Does this overlap at all with this recent Vim enhancement? 
https://github.com/vim/vim/commit/479b7e147c5e4b4c6fb387b6f705cd97fa9ba980

tarruda commented:
> Does this overlap at all with this recent Vim enhancement? 

`OptionSet` is for vim options only, I want a system that UIs can used for arbitrary options. Font, terminal keycodes, key timeout, are all UI-specific options. Vim has everything configured via it's option system. but this is not good for Neovim architecture since UIs may be developed by third parties(Imagine having to bloat neovim core with options for every existing UI).

As explained in #1913, I think the option system needs to be removed. option.c implements it's own parsing code for options when every possible configuration could be specified with plain vimscript variables and data structures, and many vim plugins already use global variables for configuration which only proves my point(If vim had an extensible option system, this would not be necessary)

Since removing option.c is not viable at the moment(not only for the amount of work, but for compatibility issues), I implemented the simplest solution I could find: Add a namespace for vimscript variables that when modified send notification to attached UIs.

justinmk commented:
> Add a namespace for vimscript variables that when modified send notification to attached UIs.

Sounds fine, but I still don't get why we pretend that UIs are the only remote clients interested in these notifications. Can `ui_attach` be renamed to `attach`? Non-UI remote clients should be able to "attach" and listen to messages. The `ui:` namespace could be renamed to `r:` (as in "remote") or `c:` (as in "client"), or something like that.

ZyX-I commented:
`ui:` dictionary is rather unexpected choice. For it the following tests are missing:

1. `ui:` evaluating to a dictionary with all options like `g:` does.
2. Testing that `call extend(ui:, {'option': value})` sends notifications.
3. Testing that `let ui:.option = value` sends notifications.
4. Testing that `let s:ui = ui: | let s:ui.option = value` sends notifications.
5. Alternative to 2.-4.: add `uinotify()` function.
6. Testing that `exists('ui:option')` works.
7. Testing that `exists('uc:option')` and `exists('u:option')` always return 0, and evaluating `uc:` and `u:` fails (`find_var_in_ht` only takes first character (`u`) in htname; I have not found how you can make it return 1 for `exists('uc:option')` with existing `ui:option`, but this test is for future scopes).

Also +1 for `r:` or `c:`: I am really not sure that I have listed everything which *may* break and needs to be tested due to two-letter scope (+1 is mainly for single letter scope).

In any case testing that `extend()` and different `let` variants work is needed: I am pretty sure some of them do not now: you are doing notification in `set_var` and not in `dict_add` or `set_var`+`dict_extend`+…

aeruder commented:
This seems to fix #2093 for me for anything I'm able to easily reproduce.

tarruda commented:
> Sounds fine, but I still don't get why we pretend that UIs are the only remote clients interested in these notifications

As I said before, the UI layer is separate from msgpack-rpc so it does need a different mechanism for notifications(which are plain callbacks set to the UI structure). So I can't use plain msgpack-rpc notifications and reuse the idea on the builtin UIs without introducing some hacky workarounds(which is something I'm trying to avoid, I want the code to be as organized as possible).

With that said, I think using msgpack-rpc to communicate with the builtin UIs could greatly simplify code organization, and since they are now running in a different thread, this could be achieved with msgpack-rpc channels over a `socketpair()`. But this would require a lot more changes(make the msgpack-rpc modules thread safe for example), so I choose the shortest path in order to prioritize a release.

>  The ui: namespace could be renamed to r: (as in "remote") or c: (as in "client"), or something like that.

It may be necessary to do this for the reasons explained by @ZyX-I(increased possibility of bugs) but it's important to understand that for neovim UIs are not necessarily remote, so "remote" and "client" are not 100% correct

tarruda commented:
> In any case testing that extend() and different let variants work is needed: I am pretty sure some of them do not now: you are doing notification in set_var and not in dict_add or set_var+dict_extend+

:+1: 

For the record, my first thought was to extend option.c to send options starting with `ui_` as notifications, but then I saw there's a lot of option parsing going on which could be messy to modify. vimscript parsing is already there, all I need is to change the functions that change the dictionary.

justinmk commented:
> the UI layer is separate from msgpack-rpc so it does need a different mechanism for notifications(which are plain callbacks set to the UI structure

I'm not suggesting that the TUI use msgpack-rpc. Currently, remote UIs call "ui_attach" and then listen for messages. I am suggesting that this be generalized to allow sending/receiving any message, not just "UI messages".

>  UIs are not necessarily remote, so "remote" and "client"

Regarding the built-in TUI as a "client" seems reasonable to me.

justinmk commented:
> Since removing option.c is not viable at the moment

When would this be viable? It is unthinkable to me. One does not fork Vim for its stellar codebase, but its ecosystem and popularity.

> I will use this in the next PR when porting the pygtk GUI, which needs to configure fonts/sizes.

Why not use `guifont` and friends for now? The `ui:` (or whatever) concept should be designed and discussed. Is it critical for 0.1? Can we instead assume that only one UI is attached at a time, until a later milestone?

fmoralesc commented:
>  The ui: (or whatever) concept should be designed and discussed. Is it critical for 0.1? Can we instead assume that only one UI is attached at a time, until a later milestone?

I agree with @justinmk in that this change would be better split out of this PR, since the other changes in it are unambiguously helpful, whereas this needs some thought and discussion.

That said, this would already be helpful to ameliorate the situation around the settings in the TUI we currently handle through environment variables, which don't map well to any already existing option (re "`guifont` and friends).

As for `option.c`: while it is a pretty messy part of the codebase, it is pretty important for backwards compatibility. I would focus in refactoring it so managing the options system is easier for maintainers instead of replacing it: a library of higher-level C functions that *helped* when dealing with strings, lists and dicts would be really appreciated, also when parsing ex commands... I also don't think making the option system dependent on viml is an improvement: it would make neovim depend on the existence of the runtime, which is an assumption we've kind of agreed not to make. In that sense, it is already bad enough that we depend on it for the clipboard management...

tarruda commented:
> When would this be viable? It is unthinkable to me. One does not fork Vim for its stellar codebase, but its ecosystem and popularity.

I have expressed myself badly, "removed" was too harsh. I just meant to say that I think that the option system is a waste due to the existence of vimscript, which could easily be enhanced to represent editor and plugin configuration. As I said, it is not viable due to compatibility since the option system existed since vi.

> Why not use guifont and friends for now? The ui: (or whatever) concept should be designed and discussed. Is it critical for 0.1? 

No, but I think it's critical that we have a builtin GUI to offer users a way to escape from the many compatibility problems related to the terminal UI. The builtin GUI will just be a port of pynvim, which is mostly complete but still misses one piece: A way to configure itself(fonts/sizes) from the user's vimrc.

Could I use `guifont` and friends? Yes, but it would require extending the "UI protocol" even further with UI-specific options. This is why I want to add standard UI configuration mechanism before 0.1, it would not only simplify the UI protocol but also allow other UIs to configure themselves.

> Can we instead assume that only one UI is attached at a time, until a later milestone?

And what would be the advantage of assuming only one UI is attached? The remote UI system is already implemented and working, all I want is to implement a cleaner way for these UIs to configure themselves.

tarruda commented:
> As for option.c: while it is a pretty messy part of the codebase, it is pretty important for backwards compatibility. I would focus in refactoring it so managing the options system is easier for maintainers instead of replacing it: a library of higher-level C functions that helped when dealing with strings, lists and dicts would be really appreciated, also when parsing ex commands... I also don't think making the option system dependent on viml is an improvement: it would make neovim depend on the existence of the runtime, which is an assumption we've kind of agreed not to make. In that sense, it is already bad enough that we depend on it for the clipboard management...

Good points, I will try to rework the last commit and integrate it into option.c, giving it some sort of extensibility for UI options.

fmoralesc commented:
>  The builtin GUI will just be a port of pynvim, which is mostly complete but still misses one piece: A way to configure itself(fonts/sizes) from the user's vimrc.

However, couldn't several clients want to have different font settings? I know with your design a GUI could just omit listening to changes to the UI settings and handle things by itself, but perhaps broadcasting by default is not necessary. 

Also: GUIs configuration files could reside at `~/.config/nvim/ui/{ID}.conf`.

tarruda commented:
> However, couldn't several clients want to have different font settings?

Yes

> I know with your design a GUI could just omit listening to changes to the UI settings and handle things by itself, but perhaps broadcasting by default is not necessary.

True, but I want to add the possibility of nvim being a remote UI of itself, and when in the "remote ui" mode, nvim will ignore UI settings send by the remote instance. I don't have all details yet(this will come after 0.1) but I will work on top of [this patch](https://github.com/neovim/neovim/issues/3119#issuecomment-130906214) and add a new function (eg: `rpcuiattach()`) which receives a channel id and makes nvim display the UI of the channel. In this mode nvim will simply ignore settings that come from the remote instance.

tarruda commented:
@justinmk I removed the ui option commit and will leave it and the builtin GUI for after 0.1. It is better to delay this instead of implementing it poorly in a rush

fmoralesc commented:
> I don't have all details yet(this will come after 0.1) but I will work on top of this patch and add a new function (eg: rpcuiattach()) which receives a channel id and makes nvim display the UI of the channel. In this mode nvim will simply ignore settings that come from the remote instance. 

Sounds super. Will this mesh with the smart ui protocol you proposed a while ago?

tarruda commented:
> Sounds super. Will this mesh with the smart ui protocol you proposed a while ago?

This is unrelated with the smart UI protocol, which only exists as a feature for third party GUIs(nvim core will only implement a simple, terminal-like GUI)

oni-link commented:
Running `test69` in `oldtest` with ASAN:
```
/home/oni-link/git/neovim/src/nvim/ui_bridge.c:238:13: runtime error: null pointer passed as argument 2, which is declared to never be null
/usr/include/string.h:47:28: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/oni-link/git/neovim/src/nvim/ui_bridge.c:238:13 in 
```

tarruda commented:
> Running test69 in oldtest with ASAN:

I wonder why travis didn't caught this. Will send a patch soon.
